hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
+
+    // Generate subgraph graph.
+    let mut subgraph_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+    let mut subgraph_succs: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+
+    for (node_id, node) in nodes.iter() {
+        if matches!(node, Node::Handoff) {
+            for &(pred, _) in new_preds[node_id].values() {
+                let pred_sg = node_subgraph[pred];
+                for &(succ, succ_idx) in new_succs[node_id].values() {
+                    if barrier_crossers.contains(&(pred, succ, succ_idx, InputBarrier::Epoch)) {
+                        continue;
                     }
-                    self.subgraph_nodes[repr_node].push(node_id);
-                    self.marked.insert(node_id);
+                    let succ_sg = node_subgraph[succ];
+                    subgraph_preds.entry(succ_sg).or_default().push(pred_sg);
+                    subgraph_succs.entry(pred_sg).or_default().push(succ_sg);
                 }
             }
+        }
+    }
 
-            let (marked, subgraph_nodes) = Default::default();
-            let mut sg_topo_sort = SubgraphTopoSort {
-                nodes: &flat_graph.nodes,
-                preds: &new_preds,
-                node_union: &mut node_union,
-                marked,
-                subgraph_nodes,
-            };
-            for node_id in flat_graph.nodes.keys() {
-                sg_topo_sort.visit(node_id);
+    let scc = {
+        // https://en.wikipedia.org/wiki/Kosaraju%27s_algorithm
+        fn visit(
+            succs: &HashMap<GraphSubgraphId, Vec<GraphSubgraphId>>,
+            u: GraphSubgraphId,
+            seen: &mut HashSet<GraphSubgraphId>,
+            stack: &mut Vec<GraphSubgraphId>,
+        ) {
+            if seen.insert(u) {
+                for &v in succs.get(&u).into_iter().flatten() {
+                    visit(succs, v, seen, stack);
+                }
+                stack.push(u);
             }
+        }
 
-            // For a `NodeId`, what `SubgraphId` does it belong to.
-            let mut node_subgraph: SecondaryMap<GraphNodeId, GraphSubgraphId> = Default::default();
-            // For a `SubgraphId`, what `NodeId`s belong to it.
-            let mut subgraph_nodes: SlotMap<GraphSubgraphId, Vec<GraphNodeId>> = Default::default();
-            // Populate above.
-            for (_repr_node, member_nodes) in sg_topo_sort.subgraph_nodes {
-                subgraph_nodes.insert_with_key(|subgraph_id| {
-                    for &node_id in member_nodes.iter() {
-                        node_subgraph.insert(node_id, subgraph_id);
-                    }
-                    member_nodes
-                });
+        let (mut seen, mut stack) = Default::default();
+        for sg in subgraph_nodes.keys() {
+            visit(&subgraph_succs, sg, &mut seen, &mut stack);
+        }
+        std::mem::drop(seen);
```

Comment:
This is a little scary. Safe? Necessary?

Reply:
Just early drops the variable, doesnt actually do much. changed to look less scary